File: basic_dynamic_programming/SAMER08D.cpp
#include<bits/stdc++.h>
using namespace std;

int main() {
    int min_sub_seq_len, i, j, k, limit;
    cin >> min_sub_seq_len;
    while (min_sub_seq_len != 0) {
        string str1, str2;
        cin >> str1;
        cin >> str2;
        str1 = " " + str1;
        str2 = " " + str2;
        int len_str1 = str1.length();
        int len_str2 = str2.length();
        int mem[len_str1][len_str2];
        memset(mem, 0, sizeof(mem[0][0]) * len_str1 * len_str2);
        for (i = 1; i < len_str1; i++) {
            for (j = 1; j < len_str2; j++) {
                if (str1[i] == str2[j]) {
                    limit = min(i, j);
                    k = 0;
                    while (k < limit && str1[i-k] == str2[j-k]) {
                        if (k + 1 >= min_sub_seq_len) {
                            mem[i][j] = max(mem[i][j], mem[i-k-1][j-k-1] + k + 1);
                        }
                        k++;
                    }
                }
                mem[i][j] = max(mem[i][j], max(mem[i-1][j], mem[i][j-1]));
            }
        }
        cout << mem[len_str1-1][len_str2-1] << endl;
        cin >> min_sub_seq_len;
    }
}

File: basic_dynamic_programming/SAMER08D.py
min_sub_seq_len = int(input())
while min_sub_seq_len != 0:
    string1 = ' '+input()
    string2 = ' '+input()
    len_string1 = len(string1)
    len_string2 = len(string2)
    mem = [[0]*len_string1 for i in range(len_string2)]
    for i in range(min_sub_seq_len, len_string2):
        for j in range(min_sub_seq_len, len_string1):
            if string2[i] == string1[j]:
                limit = min(i, j)
                k = 0
                while k < limit and string2[i-k] == string1[j-k]:
                    if k + 1 >= min_sub_seq_len:
                        mem[i][j] = max(mem[i][j], mem[i-k-1][j-k-1] + k + 1)
                    k += 1
            mem[i][j] = max(mem[i][j], mem[i-1][j], mem[i][j-1])
    print(mem[-1][-1])
    min_sub_seq_len = int(input())
